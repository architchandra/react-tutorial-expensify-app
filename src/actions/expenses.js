import uuid from 'uuid';
import { ref, push, get, remove, update } from 'firebase/database';
import database from '../firebase/firebase';



const addExpense = (expense) => ({
  type: 'ADD_EXPENSE',
  expense,
});
const startAddExpense = (expenseData = {}) => {
  return (dispatch) => {
    const {
      description = '',
      note = '',
      amount = 0,
      createdAt = 0,
    } = expenseData;
    const expense = {
      description,
      note,
      amount,
      createdAt,
    };

    return push(ref(database, 'expenses'), expense).then((ref) => {
      dispatch(addExpense({
        id: ref.key,
        ...expense,
      }));
    });
  };
};

const removeExpense = ({ id } = {}) => ({
  type: 'REMOVE_EXPENSE',
  id,
});
const startRemoveExpense = ({ id } = {}) => {
  return (dispatch) => {
    return remove(ref(database, `expenses/${id}`)).then((ref) => {
      dispatch(removeExpense({ id }));
    });
  };
};

const editExpense = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates,
});
const startEditExpense = (id, updates) => {
  return (dispatch) => {
    return update(ref(database, `expenses/${id}`), updates).then(() => {
      dispatch(editExpense(id, updates));
    });
  };
};

const setExpenses = (expenses = []) => ({
  type: 'SET_EXPENSES',
  expenses,
});
const startSetExpenses = () => {
  return (dispatch) => {
    const expenses = [];

    return get(ref(database, 'expenses')).then((snapshot) => {
      snapshot.forEach((childSnapshot) => {
        expenses.push({
          id: childSnapshot.key,
          ...childSnapshot.val(),
        });
      });
      dispatch(setExpenses(expenses));
    });
  };
};



export { startAddExpense, addExpense, startRemoveExpense, removeExpense, startEditExpense, editExpense, startSetExpenses, setExpenses };
