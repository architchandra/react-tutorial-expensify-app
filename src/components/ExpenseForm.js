import React from 'react';



class ExpenseForm extends React.Component {
  state = {
    description: '',
    amount: '',
    note: '',
  }

  onDescriptionChange = (e) => {
    const description = e.target.value;
    this.setState(() => ({ description }))
  };

  onAmountChange = (e) => {
    const amount = e.target.value;
    if(amount.match(/^\d*(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }));
    }
  };

  onNoteChange = (e) => {
    const note = e.target.value;
    this.setState(() => ({ note }));
  };
  
  render() {
    return (
      <div>
        <form>
          <input
            type="text"
            value={this.state.description}
            placeholder="Description"
            onChange={this.onDescriptionChange}
            autoFocus
          />
          <input
            type="text"
            value={this.state.amount}
            placeholder="0.00"
            onChange={this.onAmountChange}
          />
          <textarea
            value={this.state.note}
            placeholder="Note"
            onChange={this.onNoteChange}
            rows="4"
          ></textarea>
        </form>
      </div>
    );
  }
}
export default ExpenseForm;