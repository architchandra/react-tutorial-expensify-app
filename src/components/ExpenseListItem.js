import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';



const ExpenseListItem = ({id, description, note, amount, createdAt}) => {
  return (
  <Link className="list-item" to={`/edit/${id}`}>
    <div>
      {description && <h3 className="list-item__title">{description}</h3>}
      {(createdAt !== null && (typeof createdAt !== 'undefined')) && <span className="list-item__subtitle">Added on {moment(createdAt).format('Do MMMM YYYY')}</span>}
    </div>
    {amount && <h3 className="list-item__data">{numeral(amount).format('$0,0.00')}</h3>}
  </Link>
)};
export default ExpenseListItem;
