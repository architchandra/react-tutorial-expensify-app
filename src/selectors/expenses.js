import moment from 'moment';



const getExpenses = (expenses = [], { text, sortBy, startDate, endDate } = {}) => {
  if(expenses.length > 0) {
    return expenses.filter((expense) => {
      const createdAtMoment = moment(expense.createdAt);
      const hasTextFilterMatch = typeof text !== 'string' || (expense.description.toLowerCase()).includes(text.toLowerCase()) ;
      const hasStartDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true;
      const hasEndDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true;

      return hasTextFilterMatch && hasStartDateMatch && hasEndDateMatch;
    })
    .sort((expense1, expense2) => {
      if(sortBy === 'date') {
        return expense1.createdAt < expense2.createdAt ? 1 : -1;
      }
      else if(sortBy === 'amount') {
        return expense1.amount < expense2.amount ? 1 : -1;
      }
    });
  }
};



export default getExpenses;