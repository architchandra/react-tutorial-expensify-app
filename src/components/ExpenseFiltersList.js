import React from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';
import { setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate } from '../actions/filters';



class ExpenseFiltersList extends React.Component {
  state = {
    calendarFocused: null,
  };

  onDatesChange = ({ startDate, endDate }) => {
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  };
  onFocusChange = (focusedInput) => {
    this.setState(() => ({ calendarFocused: focusedInput }));
  };
  onTextChange = (e) => {
    this.props.setTextFilter(e.target.value);
  };
  onSortChange = (e) => {
    const sortByValue = e.target.value;

    if(sortByValue === 'date') {
      this.props.sortByDate();
    } else if(sortByValue === 'amount') {
      this.props.sortByAmount();
    }
  };

  render() {
    return (
      <div className="content-container">
        <div className="input-group">
          <div className="input-group__item">
            <input
              className="text-input"
              type="text"
              value={this.props.filters.text}
              onChange={this.onTextChange}
              placeholder="Search expenses"
            />
          </div>
          <div className="input-group__item">
            <select
              className="select"
              value={this.props.filters.sortBy}
              onChange={this.onSortChange}
            >
              <option value="date">Date</option>
              <option value="amount">Amount</option>
            </select>
          </div>
          <div className="input-group__item">
            <DateRangePicker
              startDate={this.props.filters.startDate}
              endDate={this.props.filters.endDate}
              onDatesChange={this.onDatesChange}
              focusedInput={this.state.calendarFocused}
              onFocusChange={this.onFocusChange}
              numberOfMonths={1}
              isOutsideRange={() => false}
              showClearDates={true}
            />
          </div>
        </div>
      </div>
    );
  }
}

const mapStateToProps = (state) => ({
  filters: state.filters,
});
const mapDispatchToProps = (dispatch) => ({
  setTextFilter: (text) => dispatch(setTextFilter(text)),
  sortByDate: () => dispatch(sortByDate()),
  sortByAmount: () => dispatch(sortByAmount()),
  setStartDate: (startDate) => dispatch(setStartDate(startDate)),
  setEndDate: (endDate) => dispatch(setEndDate(endDate)),
});

export { ExpenseFiltersList };
export default connect(mapStateToProps, mapDispatchToProps)(ExpenseFiltersList);
