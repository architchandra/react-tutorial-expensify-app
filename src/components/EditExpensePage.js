import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { editExpense } from '../actions/expenses';



const EditExpensePage = (props) => {
  return (
    <div>
      <h2>Edit an Expense</h2>
      <ExpenseForm
        expense={props.expense}
        onSubmit={(expense) => {
          props.dispatch(editExpense(props.expense.id, expense));
          props.history.push('/');
        }}
      />
    </div>
  );
};
const mapStateToProps = (state, props) => ({
  expense: state.expenses.find((expense) => expense.id === props.match.params.id),
});
export default connect(mapStateToProps)(EditExpensePage);