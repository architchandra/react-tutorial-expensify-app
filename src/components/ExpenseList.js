import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem.js';
import getExpenses from '../selectors/expenses.js';



const ExpenseList = (props) => (
  <div>
    <ul>
      {
        props.expenses.length !== 0
          ? props.expenses.map((expense, index) => <ExpenseListItem key={expense.id} {...expense} />)
          : <p>No expenses found.</p>
      }
    </ul>
  </div>
);
const mapStateToProps = (state) => ({
  expenses: getExpenses(state.expenses, state.filters)
});
export { ExpenseList };
export default connect(mapStateToProps)(ExpenseList);
