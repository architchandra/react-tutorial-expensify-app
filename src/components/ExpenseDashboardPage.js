import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseFiltersList from './ExpenseFiltersList';
import ExpensesSummary from './ExpensesSummary';



const ExpenseDashboardPage = () => (
  <div>
    <h2>Expense Dashboard Page</h2>
    <ExpensesSummary />
    <ExpenseFiltersList />
    <ExpenseList />
  </div>
);
export default ExpenseDashboardPage;
