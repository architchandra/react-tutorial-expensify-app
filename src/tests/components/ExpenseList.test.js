import React from 'react';
import { shallow } from 'enzyme';
import { ExpenseList } from '../../components/ExpenseList';
import expenses from '../fixtures/expenses';



test('should render Expense List with expenses', () => {
  const wrapper = shallow(<ExpenseList expenses={expenses} />);
  expect(wrapper).toMatchSnapshot();
});

test('should render Expense List without any expenses', () => {
  const wrapper = shallow(<ExpenseList expenses={[]} />)
  expect(wrapper).toMatchSnapshot();
});
