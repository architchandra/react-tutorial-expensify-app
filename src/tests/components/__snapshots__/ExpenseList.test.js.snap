// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Expense List with expenses 1`] = `
<div>
  <p>
    3
     expenses found
    , totalling to $520.
  </p>
  <ul>
    <ExpenseListItem
      amount={195}
      createdAt={0}
      description="Order pizza"
      id={1}
      key="1"
      note=""
    />
    <ExpenseListItem
      amount={295}
      createdAt={-345600000}
      description="Order from Pizza Hut"
      id={2}
      key="2"
      note=""
    />
    <ExpenseListItem
      amount={30}
      createdAt={345600000}
      description="Pay phone bill"
      id={3}
      key="3"
      note=""
    />
  </ul>
</div>
`;

exports[`should render Expense List without any expenses 1`] = `
<div>
  <p>
    0
     expenses found
    .
  </p>
  <ul>
    <p>
      No expenses found.
    </p>
  </ul>
</div>
`;
